### PR TITLE
feat: Add final-deletion-state Flag to Set Custom State When DeletionTimestamp is Set

### LINKED ISSUE #147: Introduce new flag to set Final-Deleting-State
**Description**
To support more complex test scenarios we need a new flag to set the State when the template-operator module is being disabled.
This topic first came up by investigation in the following issue: https://github.com/kyma-project/lifecycle-manager/issues/1080

**Acceptance Criterias**
- Add new flag to Template-Operator which allows to set the State when Module is in Deleting (DeletionTimestamp unequal zero); default Deleting (normal behaviour, but not being stuck in `Deleting`) -> Can be used to simulate more complex scenarios like the one describes in https://github.com/kyma-project/lifecycle-manager/issues/1080#issuecomment-1910427548
- Example flag name Final-Deleting-State
- Default template-operator should behave normal as before
- Cover new flag in tests
- After implementation has been done, update and unblock https://github.com/kyma-project/lifecycle-manager/issues/1080

### PR DESCRIPTION
**Pull request by LeelaChacha**
**Description**

Changes proposed in this pull request:

- Added new flag `final-deletion-state` to Template-Operator Manager - can be used to set a custom state when deletion timestamp is set (default Deleting)
- Added default state as a patch in kustomization.yaml

**Related issue(s)**

Closes #147

### CODE REVIEW

**Comment by LeelaChacha** [comment]
Note: The logic is implemented in a way that whatever, state is set, the CR will not leave that state once deletion timestamp is set. 

The other way to implement would be setting state to custom state when deletion timstamp is set and then letting the normal controller logic set the state as it is done until now. However, if we take that approach, then some states will be flickering like in [parent issue](https://github.com/kyma-project/lifecycle-manager/issues/1080). For eg: if --final-deletion-state is set to Error and deletion timestamp is set, the CR will go in state Error but then the controller would handle the error state and find no errors so it set it back to Ready until the next loop sets the CR back to Error and then the cycle repeats. Therefore, it is important that we stay in the state set in the flag. However, this means that once the flag is set and the controller is running, there is no way to change the flag, that means if a CR is set to Warning state after deletion in a test, The finalizer has to be removed manually in the test teardown, otherwise the CR will persist.

**Comment by LeelaChacha** [comment]
I repurposed the the old SampleCR tests because I would have had to create a whole new suite for test one field as messing with the reconciler in the suite would mean other tests in the suite would become flaky.

File: main.go [hunk 1]
```diff
@@ -159,6 +161,8 @@ func defineFlagVar() *FlagVar {
 		"Indicates the failure max delay in seconds")
 	flag.StringVar(&flagVar.finalState, "final-state", string(v1alpha1.StateReady),
 		"Customize final state, to mimic state behaviour like Ready, Warning")
+	flag.StringVar(&flagVar.finalDeletionState, "final-deletion-state", string(v1alpha1.StateDeleting),
+		"Customize final state when module marked for deletion, to mimic state behaviour like Ready, Warning")
```
Comment (jeremyharisch) [inline main.go:165]:
```suggestion
	flag.StringVar(&flagVar.finalDeletionState, "final-deletion-state", "",
		"Customize final state when module marked for deletion, to mimic state behaviour like Ready, Warning")
```

Default should be empty! Please also look at the other comments

File: controllers/sample_controller_rendered_resources.go [hunk 2]
```diff
@@ -112,11 +113,9 @@ func (r *SampleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
 	status := getStatusFromSample(&objectInstance)
 
-	// set state to Deleting if not set for an object with deletion timestamp
-	if !objectInstance.GetDeletionTimestamp().IsZero() &&
-		status.State != v1alpha1.StateDeleting {
-		// if the status is not yet set to deleting, also update the status
-		return ctrl.Result{}, r.setStatusForObjectInstance(ctx, &objectInstance, status.WithState(v1alpha1.StateDeleting))
+	// set state to FinalDeletionState (default is Deleting) if not set for an object with deletion timestamp
+	if !objectInstance.GetDeletionTimestamp().IsZero() && status.State != r.FinalDeletionState {
+		return ctrl.Result{}, r.setStatusForObjectInstance(ctx, &objectInstance, status.WithState(r.FinalDeletionState))
```
Comment (jeremyharisch) [inline controllers/sample_controller_rendered_resources.go:118]:
```suggestion
	if !objectInstance.GetDeletionTimestamp().IsZero() && r.FinalDeletionState != "" {
		return ctrl.Result{}, r.setStatusForObjectInstance(ctx, &objectInstance, status.WithState(r.FinalDeletionState))
```

**Review by jeremyharisch — COMMENTED** [review]
Default should not be set for a FinalDeletionState. If it is not set, then the template-operator should behave normal, without being stuck in some state. Please revisit the implementation and make needed changes to implement it in described way

**Review by jeremyharisch — APPROVED** [review]

**Pull request merged**